controllers/secretproviderclasspodstatus_controller.go

```diff
@@ -19,49 +19,136 @@ package controllers
 import (
 	"context"
 	"fmt"
-	"io/ioutil"
 	"strings"
+	"sync"
 	"time"
 
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
-	"k8s.io/apimachinery/pkg/util/wait"
+	log "github.com/sirupsen/logrus"
 
+	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
+	"sigs.k8s.io/secrets-store-csi-driver/pkg/util/fileutil"
+	"sigs.k8s.io/secrets-store-csi-driver/pkg/util/secretutil"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	log "github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
-)
-
-const (
-	certType       = "CERTIFICATE"
-	privateKeyType = "RSA PRIVATE KEY"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 // SecretProviderClassPodStatusReconciler reconciles a SecretProviderClassPodStatus object
 type SecretProviderClassPodStatusReconciler struct {
 	client.Client
+	Mutex  *sync.Mutex
 	Log    *log.Logger
 	Scheme *runtime.Scheme
 	NodeID string
 	Reader client.Reader
 	Writer client.Writer
 }
 
+func (r *SecretProviderClassPodStatusReconciler) RunPatcher(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-ticker.C:
+			if err := r.Patcher(); err != nil {
+				log.Errorf("failed to patch secret owner ref, err: %+v", err)
+			}
+		}
+	}
+}
+
+func (r *SecretProviderClassPodStatusReconciler) Patcher() error {
+	log.Debugf("patcher started")
+	r.Mutex.Lock()
+	defer r.Mutex.Unlock()
+
+	ctx := context.Background()
+	spcPodStatusList := &v1alpha1.SecretProviderClassPodStatusList{}
+	spcMap := make(map[string]v1alpha1.SecretProviderClass)
+	secretOwnerMap := make(map[types.NamespacedName][]*v1alpha1.SecretProviderClassPodStatus)
+	// get a list of all spc pod status that belong to the node
+	err := r.Reader.List(ctx, spcPodStatusList, r.ListOptionsLabelSelector())
```

Comment:
Instead of calling list all the time, can we create a spcMap cache and update the cache on spcps reconcile? Then the Patcher can rely on the spcMap cache instead.

Reply:
There is already an `spcMap` cache maintained in the patcher. The `List` call here shouldn't be expensive as it's List from the cache. Is the suggestion to maintain a cache map for the SPCPS too?

Reply:
Added a worker queue as part of this commit - https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/303/commits/a6227089818fa7714cd584b36620986947469f2c. Let me know what you think.

Reply:
> Added a worker queue as part of this commit 

Isn't the worker queue for rotation reconciler? Patcher is for reconciling ownerReference of SecretProviderClassPodStatus right?

> The List call here shouldn't be expensive as it's List from the cache

Do we want to get the latest spcPodStatusList and there spcs from the api server or is getting from the cache ok?

Reply:
Ah, the comments got mixed up. `Added a worker queue as part of this commit - a622708. Let me know what you think.` was in response to the previous comment. 

> Isn't the worker queue for rotation reconciler? Patcher is for reconciling ownerReference of SecretProviderClassPodStatus right?

Yes, that's correct. 

> Do we want to get the latest spcPodStatusList and there spcs from the api server or is getting from the cache ok?

I would consider getting from the cache here a safe operation. The cache informers get events on the watch channels as there are changes in the resources and repopulate the cache frequently. The patcher is configured with an aggressive interval of `5s` which gives enough time for the caches to be populated with the current set of resources.